Expose AuthContext in C#

jtattermusch opened the pull request:
Expose C core's auth context in C#, ready for initial review.

@jboeuf  for general sanity
@apolcyn  for C#

@nathanielmanistaatgoogle  FYI

jtattermusch commented:
@apolcyn  this should be ready for review.

apolcyn commented on src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs:
```diff
@@ -94,6 +95,42 @@ public void AuthenticatedClientAndServer()
         {
             var response = client.UnaryCall(new SimpleRequest { ResponseSize = 10 });
             Assert.AreEqual(10, response.Payload.Body.Length);
+
```
nit: extra blank line here

jtattermusch replied:
done.

apolcyn commented on src/csharp/Grpc.Core/AuthProperty.cs:
```diff
@@ -0,0 +1,125 @@
+        }
+
+        /// <summary>
+        /// Gets the string of value of the property.
```
nit: `s/"string of value"/"string value"`

jtattermusch replied:
done

apolcyn commented on src/csharp/Grpc.Core/AuthProperty.cs:
```diff
@@ -0,0 +1,125 @@
+        }
+
+        /// <summary>
+        /// Creates and instance of <c>AuthProperty</c>.
```
nit: s/"and instance"/"an instance"`

jtattermusch replied:
done

apolcyn commented on src/csharp/Grpc.Core/Internal/AuthContextSafeHandle.cs:
```diff
@@ -0,0 +1,121 @@
+            {
+                var authProperty = PtrToAuthProperty(authPropertyPtr);
+
+                List<AuthProperty> properties = null;
```
nit: IMO only but maybe simpler as:
```
                if (!propertiesDict.ContainsKey(authProperty.Name))
                {
                    propertiesDict[authProperty.Name] = new List<AuthProperty>();
                }
                propertiesDict[authProperty.Name].Add(authProperty);
```

jtattermusch replied:
done.

apolcyn commented on src/csharp/Grpc.Core/Internal/MarshalUtils.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        /// <summary>
+        /// Get string from a ASCII encoding byte array.
```
small wording nits on summary comments. s/a/an

jtattermusch replied:
done

apolcyn commented on src/csharp/Grpc.Core/AuthProperty.cs:
```diff
@@ -0,0 +1,125 @@
+        }
+
+        /// <summary>
+        /// Gets the string value of the property.
```
always safe to assume the value read out is UTF8 encoded?

jtattermusch replied:
I think so.

jtattermusch replied:
To me that's an implementation detail. Normally, the auth property will be created by gRPC and users will just read string values and from that perspective, the original encoding is not really important.

apolcyn commented on src/csharp/Grpc.Core/AuthProperty.cs:
```diff
@@ -0,0 +1,125 @@
+        /// </summary>
+        /// <param name="name">the name</param>
+        /// <param name="valueBytes">the binary value of the property</param>
+        public static AuthProperty Create(string name, byte[] valueBytes)
```
since it looks like AuthProperty only created within grpc and passed up to the app, does the `Create` method here need to be exposed?

jtattermusch replied:
It is useful for testing, otherwise there is no way for the users to create instances of AuthProperty.

apolcyn replied:
For that, need the public constructor for `AuthContext` too?

jtattermusch replied:
I'd better not expose the public constructor for AuthContext for now. Once a constructor is published, it's hard to change the internals of the auth context if changes are necessary. We can later add the constructor once the existing implementation proves to be working well for people.

apolcyn commented on src/csharp/Grpc.Core/AuthContext.cs:
```diff
@@ -0,0 +1,127 @@
+    /// AuthContext is the only reliable source of truth when it comes to authenticating calls.
+    /// Using any other call/context properties for authentication purposes is wrong and inherently unsafe.
+    /// </summary>
+    public class AuthContext
```
more idiomatic to use the `Enumerable<AuthProperty>`?

one thought on API is it seems `AuthContext` could be more like a dictionary.

basically `Dictionary<string, List<Property>>` could be more like `Dictionary<string, List<Value>>`?

jtattermusch replied:
yeah, ideally properties should be a read only multi-map, but unfortunately C# libraries don't have a type for multimap nor they have the right type that would allow use to preserve immutability. So what I'm exposing is my best attempt to make the API clear and preserve all the properties I want it to have.

jtattermusch commented:
I believe I've addressed/answered all comments.
Also rebased against the latest master and added a note that AuthContext/AuthProperty api is experimental for now (so we have a bit of freedom to make adjustments to the API if needed).
Btw, master now has a high-level doc about server side auth (docs/server_side_auth.md)

PTAL.

jtattermusch commented:
Known failures:
- https://github.com/grpc/grpc/issues/9399
- https://github.com/grpc/grpc/issues/9956

jtattermusch commented:
@dgquintas  FYI I'd like to merge this one before cutting the 1.2.x release branch.

dgquintas commented:
@jtattermusch in that case add the "release blocker" label
